### ISSUE TITLE
You must feed a value for placeholder tensor 'input_example_tensor' with dtype string and shape [?]

### ISSUE THREAD

**Issue by yejw5** [tensorflow/serving#1404]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Centos 7.6.1810
- **TensorFlow Serving installed from (source or binary)**: source
- **TensorFlow Serving version**: 1.2.0

### Describe the problem
**Firstly, export serving model as follows:**

There is a features map for SparseTensor used as inputs, get by:
```
# func init_idmapping() returns the keys list for features
def init_spec():
    features = {}
    for i in init_idmapping():
        features[str(i)] = tf.VarLenFeature(tf.int64)
    return features
```

And than generate an `input_receiver` as follows:
```
feature_spec = init_spec()
input_receiver = tf.estimator.export.build_parsing_serving_input_receiver_fn(feature_spec)
```
Export serving model:
```
with tf.Session() as sess:
  model_path = FLAGS.model_path + "/checkpoint/"
  serving_path = FLAGS.serving_path + "/2"

  latest_checkpoint_file_path = tf.train.latest_checkpoint(model_path)
  print(latest_checkpoint_file_path)
  saver.restore(sess, latest_checkpoint_file_path)

  # define output path
  print("DNN recall model restored")
  export_path = serving_path

  # build graph
  builder = tf.saved_model.builder.SavedModelBuilder(export_path)
  export_outputs = {"predict": export_output_lib.PredictOutput(pred)}

  # print "receiver_tensors:"
  # print input_receiver().receiver_tensors
  # output graph
  signature_def_map = export_helpers.build_all_signature_defs(
      receiver_tensors=input_receiver().receiver_tensors,
      export_outputs=export_outputs,
      receiver_tensors_alternatives=getattr(input_receiver(),
                                            'receiver_tensors_alternatives',
                                            None),
      serving_only=True)

  # print "signature_def_map:"
  print signature_def_map
  legacy_init_op = tf.group(tf.tables_initializer(), name='legacy_init_op')
  builder.add_meta_graph_and_variables(
      sess=sess,
      tags=[tf.saved_model.tag_constants.SERVING],
      signature_def_map=signature_def_map,
      assets_collection=ops.get_collection(ops.GraphKeys.ASSET_FILEPATHS),
      strip_default_attrs=False,
      legacy_init_op=legacy_init_op,
      saver=saver)
  builder.save()
```
And the printed `signature_def_map` is:
```
{'predict': inputs {
  key: "examples"
  value {
    name: "input_example_tensor_1:0"
    dtype: DT_STRING
    tensor_shape {
      dim {
        size: -1
      }
    }
  }
}
outputs {
  key: "output"
  value {
    name: "add_1:0"
    dtype: DT_FLOAT
    tensor_shape {
      dim {
        size: -1
      }
      dim {
        size: 1
      }
    }
  }
}
method_name: "tensorflow/serving/predict"
}
```

**Secondly, predict with random data in client:**
```
# func init_idmapping() returns the keys list for features
id_mapping = init_idmapping()
id_spec = init_spec()

feature_map = {}
for key in id_mapping:
  data = np.random.randint(10000,
                           size=(int(id_mapping[key]))).astype(np.int64)
  feature_map[key] = tf.train.Feature(int64_list=tf.train.Int64List(
      value=data))
features = tf.train.Features(feature=feature_map)
example = tf.train.Example(features=features)
serialized = example.SerializeToString()
request.inputs['examples'].CopyFrom(
    tf.contrib.util.make_tensor_proto(serialized, shape=[1], dtype=tf.string))
result_counter.throttle()
result_future = stub.Predict.future(request, 5.0)  # 5 seconds
result_future.add_done_callback(_create_rpc_callback(result_counter))
```
**But get error response:**
```
<_Rendezvous of RPC that terminated with:
        status = StatusCode.INVALID_ARGUMENT
        details = "You must feed a value for placeholder tensor 'input_example_tensor' with dtype string and shape [?]
         [[{{node input_example_tensor}}]]"
        debug_error_string = "{"created":"@1564388367.758266250","description":"Error received from peer","file":"src/core/lib/surface/call.cc","file_line":1017,"grpc_message":"You must feed a value for placeholder tensor 'input_example_tensor' with dtype string and shape [?]\n\t [[{{node input_example_tensor}}]]","grpc_status":3}"
>

```

### Conclusion
From the printed `signature_def_map` log, it seems that what I should do is transfering a tensor with name "examples" and a serialized Tensor for predicting. But I still got error for lack of `placeholder tensor 'input_example_tensor'`. Could you provide some suggestions?

**Comment by rmothukuru** [comment]
@yejw5 ,
In your code, you are using the function, `tf.estimator.export.build_parsing_serving_input_receiver_fn`. As per the code mentioned in [this link](https://www.tensorflow.org/guide/saved_model#prepare_serving_inputs), under the hood behavior of `tf.estimator.export.build_parsing_serving_input_receiver_fn` is mentioned below:

```
def serving_input_receiver_fn():
  """An input receiver that expects a serialized tf.Example."""
  serialized_tf_example = tf.placeholder(dtype=tf.string, shape=[default_batch_size],
                                         name='input_example_tensor')
  receiver_tensors = {'examples': serialized_tf_example}
  features = tf.parse_example(serialized_tf_example, feature_spec)
  return tf.estimator.export.ServingInputReceiver(features, receiver_tensors)
```
That might be the reason the Input is expecting a Placeholder. 

TLDR, to conclude, you can either use `tf.estimator.export.build_raw_serving_input_receiver_fn`  or prepare Serving Data using SavedModelBuilder as given in this [Example](https://github.com/tensorflow/serving/blob/master/tensorflow_serving/servables/tensorflow/testdata/saved_model_half_plus_two.py) because it is a very lengthy procedure to perform Inference using Predict API with Input Data Type as String.

**Comment by yejw5** [comment]
@rmothukuru Thank you for reply.

The problem is resolved. We have saved checkpoint file of trained model, and then restore graph to saving as serving format pb. In this procedure, `tf.estimator.export.build_parsing_serving_input_receiver_fn` was called repeatedly (the training code also called parse_example func and create placeholder `input_example_tensor`), a new 
 placeholder named `input_example_tensor_1` created, so setting the signature with `input_example_tensor_1`. So in serving predict, `input_example_tensor` cannot be found...

And thanks for your suggestions, we will try the approaches. Many thanks!

**Comment by rmothukuru** [comment]
Are you satisfied with the resolution of your issue?<br> [Yes](https://docs.google.com/forms/d/e/1FAIpQLSfaP12TRhd9xSxjXZjcZFNXPGk4kc1-qMdv3gc6bEP90vY1ew/viewform?entry.85265664=Yes&entry.2137816233=1404)<br> [No](https://docs.google.com/forms/d/e/1FAIpQLSfaP12TRhd9xSxjXZjcZFNXPGk4kc1-qMdv3gc6bEP90vY1ew/viewform?entry.85265664=No&entry.2137816233=1404)

**Comment by yejw5** [comment]
Yes！ And I have finished the questionnaire. Thank you!

**Comment by rmothukuru** [comment]
Thank you!
